### PR TITLE
Introduce DirectoryObject and FileObject in Workflow

### LINKED ISSUE #702: Concepts and ideas for a node-based pyiron applied to file-based executables
I have played a bit with combining our new ideas regarding a node-based pyiron with handling files for codes such as Lammps that rely on file input and output. The example works, but I have many ideas how to extend and improve it. To keep it simple I didn't
 try to extrect single self-contained pyiron functions (which we should do) but used the full pyiron stack. 

Below is the Markdown version of my JupyterNotebook :

# Toy implementation and first tests of a File Object Class for workflows

```python
from pathlib import Path
import os
```

### Create file data type/object

```python
import os
from pathlib import Path

class DirectoryObject:
    def __init__(self, directory):
        self.directory = Path(directory)
        self.create()

    def create(self):
        if not self.directory.exists():
            self.directory.mkdir(parents=True)
            print(f"Directory '{self.directory}' created successfully.")
        else:
            print(f"Directory '{self.directory}' already exists.")

    def delete(self):
        if self.directory.exists():
            # Remove all files within the directory
            for file in os.listdir(self.directory):
                file_path = self.directory / file
                print (file, file_path.is_file(), file_path.is_dir())
                if file_path.is_file():
                    file_path.unlink()
                    print(f"File '{file_path}' deleted successfully.")
            
            self.directory.rmdir()
            print(f"Directory '{self.directory}' deleted successfully.")
        else:
            print(f"Directory '{self.directory}' does not exist.")

    def list_files(self):
        if self.directory.exists():
            files = os.listdir(self.directory)
            if files:
                print(f"Files in directory '{self.directory}':")
                for file in files:
                    print(file)
            else:
                print(f"No files found in directory '{self.directory}'.")
        else:
            print(f"Directory '{self.directory}' does not exist.")
            
    def __len__(self):
        files = []
        if self.directory.exists():
            files = os.listdir(self.directory)
        return len(files)
            
    def __repr__(self):
            return f"DirectoryObject(directory='{self.directory}' with {len(self)} files)"            
```

```python
# Example usage
directory_handler = DirectoryObject('WorkingDir_new')

directory_handler.list_files()
```

```python
directory_handler.delete()
directory_handler.list_files()
directory_handler
```

### Create file data type/object

```python
class FileObject:
    def __init__(self, path):
        self.path = Path(path)
        
    def __repr__(self):
        if not self.path.is_file():
            return 'File does not exist'
        f_str = f'File object(path=\'{self.path}\') {os.linesep}'
        with self.path.open(mode='r') as f:
            f_str += f.read()
        return f_str    
        
    def write(self, text, mode='w'):
        with self.path.open(mode=mode) as f:
            f.write(text)
            
    def write_text(self, text, mode='w'):
        self.path.write_text(text)  
            
    def delete(self): 
        self.path.unlink()
        
    def move_to_dir(self, target):
        if isinstance(target, DirectoryObject):
            target = target.directory
        else:    
            target = Path(target)
        source = target / self.path.name

        if target.is_dir():
            self.path.rename(source)
        self.path = source 
        
    def rename(self, new_filename):
        new_path = self.path.parent / new_filename
        if new_path.is_file():
            raise ValueError('File exists already')
        self.path.rename(new_path)
        self.path = new_path
```

#### Application examples

```python
work_dir = DirectoryObject('WorkingDir')

fo = FileObject('./input.test')
fo.write('my first test')
fo
```

```python
fo_new = DirectoryObject('WorkingDir_new')
fo.move_to_dir(fo_new)
#fo.move_to_dir('WorkingDir')
fo_new, fo
```

```python
fo.delete()
```

```python
fo_new.delete()
```

#### ListFileObject

```python
class ListFileObject:
    def __init__(self, file_list=[], directory='.'):
        self.files = {}
        for file in file_list:
            self.files[file] = FileObject(Path(directory) / Path(file) )
            
    def __repr__(self):
        return f'ListFileObject({list(self.files.keys())})'
            
```

### Create example workflow for Lammps

```python
from string import Template
import dask
```

#### Create some of the Lammps files using pyiron

Note: For the toy model I use pyiron to generate the structure.inp and potential files. For a more practical application/implementation this should be replaced by nodes that directly translate e.g. the pyiron atomic structure into a Lammps input.ini file.

```python
class WorkflowResources:
    def __init__(self, working_directory, executable, job_name, project):
        self.working_directory = working_directory
        self.executable = executable
        self.job_name = job_name
        self.project = project

@dask.delayed  # does not work with dask
def lammps_setup(structure, project_path='.', job_name='job'):
    from pyiron import Project

    pr = Project(project_path)

    lammps = pr.create.job.Lammps(job_name)
    lammps.structure = structure
    list_pot = lammps.list_potentials()
    lammps.potential = list_pot[0]
    lammps.write_input()
    return WorkflowResources(working_directory=lammps.working_directory, 
                     executable=lammps.executable, 
                     job_name=job_name, 
                     project=pr)
```

Note: The following statement fails in dask when implementing multiple returns (can then be only resolved via getitem)

    working_dir, executable = lammps_setup(structure=bulk, label=pr.name)
    
It is therefore better to introduce a complex data object    

```python
control_inp_MD_str = Template('''
units metal
dimension 3
boundary p p p
atom_style atomic
read_data structure.inp
include potential.inp
fix ensemble all nvt temp $temperature $temperature 0.1
variable dumptime  equal $n_print 
variable thermotime  equal $n_print 
timestep 0.001
velocity all create $temperature_x2 67525 dist gaussian
dump 1 all custom $${dumptime} dump.out id type xsu ysu zsu fx fy fz vx vy vz
dump_modify 1 sort id format line "%d %d %20.15g %20.15g %20.15g %20.15g %20.15g %20.15g %20.15g %20.15g %20.15g"
thermo_style custom step temp pe etotal pxx pxy pxz pyy pyz pzz vol
thermo_modify format float %20.15g
thermo $${thermotime}
run $n_ionic_steps
''')

# print(control_inp_MD_str.substitute(temperature=900, temperature_x2=2*900, n_print=100, n_ionic_steps=1000))
```

```python
@dask.delayed
def LammpsMD_control(temperature=900, n_print=100, n_ionic_steps=1000, directory='.'):
    
    text = control_inp_MD_str.substitute(temperature=temperature, 
                                         temperature_x2=2*temperature, 
                                         n_print=n_print, 
                                         n_ionic_steps=n_ionic_steps)
    
    fo = FileObject(Path(directory) / 'control.inp')
    fo.write_text(text)
    return fo    
```

```python
@dask.delayed
def LammpsControlFile(directory='.', 
                      text=None):
    fo = FileObject(Path(directory) / 'control.inp')
    fo.write_text(text)
    return fo    
```

```python
@dask.delayed
def runLammps(control,
              directory='.', 
              executable=None):
    import subprocess

    subprocess.run(f'cd {directory}; {str(executable)} > error.out', shell=True)
    
    return ListFileObject(['dump.out', 'error.out', 'log.lammps'], 
                          directory=directory)
```

```python
@dask.delayed
def lammps_collect_dump(dump_file:FileObject):
    import numpy as np
    
    data_lst = []
    
    n_atoms = 4
    n_shift = 9 + n_atoms    

    with open(dump_file.path, 'r') as f:
        for i in range(8):
            line = f.readline()
            if i==3:
                n_atoms = int(line.split()[-1])
                n_shift = 9 + n_atoms
                print ('atoms:', n_atoms)         
        for i, line in enumerate(f.readlines()):
            if (i%n_shift > 0) & (i%n_shift<n_shift-8):
                data_lst.append([float(number) for number in line.split()[2:]])
                
        data_lst = np.array(data_lst).T

    label_lst = 'xsu ysu zsu fx fy fz vx vy vz'.split()
    out_dict = {}
    out_dict['unwrapped_positions'] = data_lst[:3].reshape([-1, n_atoms, 3])
    out_dict['forces'] = data_lst[3:6].reshape([-1, n_atoms, 3])
    return out_dict
```

```python
@dask.delayed
def plot_tensor_rank3(tensor, axis=0, ylabel=''):
    import matplotlib.pylab as plt
    
    plt.ylabel(ylabel)
    return plt.plot(tensor[:,:,axis])
```

### Construct and run workflow

```python
from pyiron import Project

pr = Project('test')
bulk = pr.create.structure.bulk('Al', cubic=True)
```

```python
%%time
resources = lammps_setup(structure=bulk, project_path=pr.name, job_name='lammps')

control = LammpsMD_control(temperature=20, 
                           n_ionic_steps=1000, 
                           directory=resources.working_directory)
lammps = runLammps(control, 
                   directory=resources.working_directory,
                   executable=resources.executable)

out_dict = lammps_collect_dump(lammps.files['dump.out'])

plot = plot_tensor_rank3(out_dict['unwrapped_positions'], axis=2, ylabel='Positions');
plot.compute();
```

The above warning occurs in setup_lammps and is likely related to not closing the database connection in our pyiron job object. Check how this could be done.

### Visualize workflow graph

```python
plot.dask
```

```python
plot.visualize(engine="cytoscape")
```

Missing in dask/present implementation:
- access to input after initialization
- construction of explicit workflows
- what is a good syntax to access input of nodes, subnotes etc. in a workflow

### PR DESCRIPTION
**Pull request by samwaseda**
This PR allows for working files which some softwares require.

- `FileObject` handles only the file and has no information of address i.e. it must be attached to a `DirectoryObject`
- The central path is known to `Workflow` -> a `Node` cannot create a directory if it is not attached to a `Workflow`
- The directory is created only if `workflow.working_directory` is called.
- Path: `workflow_name/function_name/file_name`

Example I:

```python
from pyiron_contrib.workflow import Workflow
from pyiron_contrib.workflow.node import Node

def f(x):
    return x

wf = Workflow("TEST")
n_f = Node(f, "f")
wf.add(n_f)
my_file = n_f.working_directory.create_file('new_file')  # Creates all directories
my_file.write('something')

with open(my_file.path, 'r') as f:
    print(f.read())

wf.working_directory.delete()  # deletes everything
```

Output:

```
TEST/f/new_file
something
```

Example II

```python
def f(x):
    return x

n_f = Node(f, "f")
my_file = n_f.create_file('new_file')  # Error, because not attached to a workflow
```

Closes #702

### CODE REVIEW

**Review by liamhuber — CHANGES_REQUESTED** [review]
lgtm. One UX question.

Needs tests though. `files.py` is super readable and I don't see anything that could go wrong, but tests should be there at least for the coverage 😉 For the workflow and nodes I actually see tests being useful as they lock-in two promises:
- Paths are based off labels (this is actually mis-spoken in the PR comment, it uses the node label not the function name -- these are the same by default but can change, and even do change in your example!)
- Nodes not being able to create a file if not in a workflow (maybe we want to change that spec later, but for now I like it)

At first I thought it might be overkill to put `create_file` right on `Node` (as it wouldn't be accessible to any node that doesn't define `self` in the function argument), but actually (a) I don't mind the extra code that much, and (b) maybe we can exploit it when it comes time to serialize IO. So I want to keep in the back of our minds the idea of refactoring these methods away to some more specific child class, but for now it seems perfectly appropriate to have it here.

**Comment by samwaseda** [comment]
> Needs tests though.

Ah yes, I left a comment in the [botched previous PR](https://github.com/pyiron/pyiron_contrib/pull/722) saying I was waiting for the first comments but I forgot to write the same in this PR.


> * Paths are based off labels (this is actually mis-spoken in the PR comment, it uses the node label not the function name -- these are the same by default but can change, and even do change in your example!)

Yeah it's good that we talk about it, because until I saw [this PR](https://github.com/pyiron/pyiron_contrib/pull/727) I didn't really think that the label can be different.

> * Nodes not being able to create a file if not in a workflow (maybe we want to change that spec later, but for now I like it)
> 
> At first I thought it might be overkill to put `create_file` right on `Node` (as it wouldn't be accessible to any node that doesn't define `self` in the function argument), but actually (a) I don't mind the extra code that much, and (b) maybe we can exploit it when it comes time to serialize IO. So I want to keep in the back of our minds the idea of refactoring these methods away to some more specific child class, but for now it seems perfectly appropriate to have it here.

I also thought about defining it in `node.working_directory.create_file`, and to be honest I don't have a very strong feeling, so I can also move it there.

File: pyiron_contrib/workflow/files.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        path = self.path / Path(file_name)
+        with path.open(mode=mode) as f:
+            f.write(content)
+
+    def create_subdirectory(self, path):
+        return DirectoryObject(self.path / path)
+
+
+class FileObject:
+    def __init__(self, file_name: str, directory: DirectoryObject):
```
[the remark this answers is not in the document]
Comment (samwaseda) [reply pyiron_contrib/workflow/files.py:45]: yeah that might be a good idea

[the remark this answers is not in the document]
Comment (samwaseda) [reply pyiron_contrib/workflow/files.py:45]: although I was thinking about using the same object for reading as well, i.e. for all files which are created during `run` should also be accessed with this object, so I'm gonna put the warning in `write`.

**Comment by samwaseda** [comment]
> A simple example is a Murnaghan node, where several atomistic jobs (with each a separate working directory) need to be created. Thus, nodes should be able to create subdirectories with the working directory of the workflow being the parent directory.

There, I think there's a bit of a difference in terminology with respect to the current implementation of `Workflow`: a `Node` is the smallest unit that wraps **one** function, while a `Workflow` can contain multiple nodes and itself can act like a node. This means, in the example of Murnaghan (which is a bit difficult to talk about, because parallel jobs are not fully conceptualised on our side), each of the atomistic job is a `Node`, but the Murnaghan job itself is a `Workflow`. So I think it would be sufficient to have one directory for one node.

This being said, there indeed should be a functionality that allows a workflow folder within a parent workflow. I just didn't do it because I didn't see a possibility in the current implementation how to see whether a workflow is attached to a parent workflow (i.e. I don't know if it's possible for a child workflow to know whether it has a parent workflow).

Comments from @liamhuber might be appreciated here.

**Comment by liamhuber** [comment]
> but the Murnaghan job itself is a `Workflow`

No, Joerg is correct; you can in principle have a node that executes the entire Murnaghan procedure. It doesn't even need to make sub-nodes out of the steps! The idea that nodes are functional and idempotent is a "best practice" and not strictly enforced, and even here you could have a node that starts a pyiron project, makes a bunch of lammps jobs, runs everything, scrapes the results, and finishes that _is_ both idempotent and functional! Of course, you also could make a `Murnaghan` workflow where each calculation is a node, and then you don't need this.

A `task` is just a computational unit and a `Node.run()` might generate multiple of them at once. There's no reason to forbid a task from having its own working directory. This is already possible from inside the node of course, you can use `self.working_directory` as a root and then make new sub-dirs as part of the task. I don't have any objection to adding tools that make this subdirectory creation easier.

**Comment by samwaseda** [comment]
Ok then it doesn't sound like something systematic, so I propose the following solution (which is only marginally different from the current implementation):

```python
def my_function(x):
    return x

wf = Workflow("TEST")
n_f = Node(my_function, "output")
wf.add(n_f)
node_file = n_f.working_directory.create_file('node_directory_file')
node_file.write('This one is written in `my_function` and is inside the node directory')
sub_directory = n_f.working_directory.create_subdirectory("node_subdirectory")
node_sub_file = sub_directory.create_file('node_subdirectory_file')
node_sub_file.write('This one is written in `node_subdirectory` and is inside the node subdirectory')
```

```python
print(n_f.working_directory.list_content())
```

Output:

```python
{'dir': ['TEST/my_function/node_subdirectory'],
 'file': ['TEST/my_function/node_directory_file'],
 'mount': [],
 'symlink': [],
 'block_device': [],
 'char_device': [],
 'fifo': [],
 'socket': []}
```

```python
print(node_file.path)
```

Output:

```python
TEST/my_function/node_directory_file
```

```python
print(sub_directory.list_content())
```

Output:

```python
{'dir': [],
 'file': ['TEST/my_function/node_subdirectory/node_subdirectory_file'],
 'mount': [],
 'symlink': [],
 'block_device': [],
 'char_device': [],
 'fifo': [],
 'socket': []}
```

```python
print(node_sub_file.path)
```

Output:

```python
TEST/my_function/node_subdirectory/node_subdirectory_file
```

One thing that changes here is that `create_file` is not part of `Node` anymore, which comes back to the point that @liamhuber had made. Would this be acceptable?

**Review by liamhuber — COMMENTED** [review]
A minor question and a readability nit, but otherwise lgtm now.

Regarding the larger question brought up by Joerg, I think the current implementation where neither workflow nor node have special creation routines but _need_ to go via their `working_directory` is a good solution -- all the capability is available, and we're not biasing towards one particular usage style.

File: pyiron_contrib/workflow/files.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    def file_name(self):
+        return self._file_name
+
+    @property
+    def path(self):
+        return self.directory.path / Path(self._file_name)
+
+    def write(self, content, mode='w'):
+        if self.is_file():
+            warnings.warn(f"{self.file_name} already exists")
```
[the remark this answers is not in the document]
Comment (samwaseda) [reply pyiron_contrib/workflow/files.py]: I changed the default mode.

File: tests/unit/workflow/test_workflow.py [hunk 3]
```diff
@@ -107,3 +107,11 @@ def plus_one(x: int = 0) -> int:
             return x + 1
 
         self.assertEqual(plus_one().outputs.y.value, 1)
+
+    def test_working_directory(self):
+        wf = Workflow("wf")
+        self.assertTrue(wf._working_directory is None)
+
```
Comment (liamhuber) [inline tests/unit/workflow/test_workflow.py:114]:
```suggestion
self.assertIsInstance(wf.working_directory, DirectoryObject)
self.assertTrue(wf.working_directory.path.absolute().endswith(wf.label)

```

Maybe the endswith needs a "/" or something? Also I guess we need similar tests for `Node`, where it starts as none, raises an error when asked for, and then passes the same endswith test after a parent is added

**Review by liamhuber — COMMENTED** [review]
Nice. A couple minor suggestions for extending the tests, but otherwise lgtm

**Comment by liamhuber** [comment]
I could have sworn GitHub used to rebase stacked PRs onto main _automatically_ when their target got merged and deleted....

**Comment by liamhuber** [comment]
Test failures on `` look like they're just typical windows stuff, although the `is_mount` thing is new to me:

```
======================================================================
ERROR: test_write (workflow.test_files.TestFiles)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\pyiron_contrib\pyiron_contrib\tests\unit\workflow\test_files.py", line 21, in test_write
    self.assertTrue("test/test.txt" in self.directory.list_content()['file'])
  File "D:\a\pyiron_contrib\pyiron_contrib\pyiron_contrib\workflow\files.py", line 48, in list_content
    return categorize_folder_items(self.path)
  File "D:\a\pyiron_contrib\pyiron_contrib\pyiron_contrib\workflow\files.py", line 31, in categorize_folder_items
    if getattr(item, f"is_{tt}")():
  File "C:\Miniconda3\envs\my-env\lib\pathlib.py", line 1461, in is_mount
    raise NotImplementedError("Path.is_mount() is unsupported on this system")
NotImplementedError: Path.is_mount() is unsupported on this system

======================================================================
FAIL: test_path (workflow.test_files.TestFiles)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\pyiron_contrib\pyiron_contrib\tests\unit\workflow\test_files.py", line 30, in test_path
    self.assertEqual(str(f.path), "test/test.txt")
AssertionError: 'test\\test.txt' != 'test/test.txt'
- test\test.txt
?     ^
+ test/test.txt
?     ^


----------------------------------------------------------------------
```

**Review by liamhuber — COMMENTED** [review]
It would be nice to extend the tests as [here](https://github.com/pyiron/pyiron_contrib/pull/725/files#r1238726659), but otherwise once the tests are passing this is looking alright to me.

**Comment by samwaseda** [comment]
> Test failures on `` look like they're just typical windows stuff, although the `is_mount` thing is new to me

I was unnecessarily calling all functions starting with `is`. Now it should be fine.

**Review by liamhuber — APPROVED** [review]
Super.

I think there is a way to more elegantly convert `"test/tests"` to look like a path from the current system than to convert all the references to "/", but what you do here is clear and effective so it's also ok with me.
